atc/db/version_reaper.go

```diff
@@ -0,0 +1,124 @@
+package db
+
+import (
+	"code.cloudfoundry.org/lager"
+	"context"
+	"database/sql"
+	sq "github.com/Masterminds/squirrel"
+)
+
+type resourceVersionReaper struct {
+	conn                   Conn
+	logger                 lager.Logger
+	maxVersionsPerResource int
+}
+
+func NewResourceVersionReaper(logger lager.Logger, conn Conn, maxVersionsPerResource int) *resourceVersionReaper {
+	return &resourceVersionReaper{
+		conn:                   conn,
+		logger:                 logger,
+		maxVersionsPerResource: maxVersionsPerResource,
+	}
+}
+
+func (r *resourceVersionReaper) ReapVersions(ctx context.Context) error {
+	rcsIds, err := r.getResourceConfigScopesToReap(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, rcsId := range rcsIds {
+		checkOrder, err := r.getLastCheckOrderToRetain(ctx, rcsId)
+		if err != nil {
+			return nil
+		}
+
+		if checkOrder == 0 {
+			continue
+		}
+
+		n, err := r.deleteVersions(ctx, rcsId, checkOrder)
+		if err != nil {
+			r.logger.Error("failed-to-delete-verions", err, lager.Data{"resource-config-scope-id": rcsId})
+			continue
+		}
+		r.logger.Info("deleted-versions", lager.Data{"resource-config-scope-id": rcsId, "deleted-versions": n})
+	}
+
+	return nil
+}
+
+// getResourceConfigScopesToReap return a list of resource_config_scope_id who
+// has more than maxVersionsPerResource versions.
+func (r *resourceVersionReaper) getResourceConfigScopesToReap(ctx context.Context) ([]int, error) {
+	rows, err := psql.Select("count(*) a", "resource_config_scope_id").
+		From("resource_config_versions").
+		Where(sq.NotEq{
+			"check_order": 0,
+		}).
```

Comment:
I think we need a `GROUP BY resource_config_scope_id` to allow this query to work. But also when I ran this query on an environment that has a lot of resource versions, it took 12 seconds to run. 

```
Sort  (cost=135253.56..135255.41 rows=742 width=12) (actual time=12913.111..12913.138 rows=434 loops=1)
  Sort Key: (count(*)) DESC
  Sort Method: quicksort  Memory: 45kB
  ->  HashAggregate  (cost=135210.76..135218.18 rows=742 width=12) (actual time=12912.473..12913.007 rows=434 loops=1)
        Group Key: resource_config_scope_id
        Filter: (count(*) > 100)
        Rows Removed by Filter: 2363
        ->  Seq Scan on resource_config_versions  (cost=0.00..118432.66 rows=2237080 width=4) (actual time=0.633..12211.803 rows=2239886 loops=1)
              Filter: (check_order <> 0)
              Rows Removed by Filter: 1
Planning time: 0.116 ms
Execution time: 12913.192 ms
```

You might be able to add an index to `check_order` in order to speed it up? I'm not too sure though and I'm also hesitant to add that index to this table (if there is another option).

Reply:
Good catch! My original SQL statement is:

```sql
select count(*) a, resource_config_scope_id from resource_config_versions where check_order != 0 group by resource_config_scope_id having count(*) > 1000 order by a desc
```

In my DB, table `resource_config_versions` has 3.3M tuples, and the query takes less than 2 seconds. I also tried to get rid of `check_order != 0` from the query, it didn't speed up significantly. But considering the following 2 statements also need to sort on `check_order`, I think it's worth adding an index on `resource_config_scope_id+check_order`. I guess the gain of keep `resource_config_versions` small would be bigger than the cost of adding a new index.